dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/cloudera/manager/AbstractClouderaYarnApplicationTask.java

```diff
@@ -0,0 +1,131 @@
+
+public abstract class AbstractClouderaYarnApplicationTask extends AbstractClouderaManagerTask {
+  private static final DateTimeFormatter isoDateTimeFormatter =
+      DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
+
+  private final String fromDate;
+
+  public AbstractClouderaYarnApplicationTask(String fileName, int includedLastDays) {
+    super(String.format("%s-%dd.jsonl", fileName, includedLastDays));
+    Preconditions.checkArgument(
+        includedLastDays > 1,
+        String.format("Amount of days must be a positive number. Get %d.", includedLastDays));
+    this.fromDate = buildISODateTime(includedLastDays);
+  }
+
+  private String buildISODateTime(int deltaInDays) {
+    ZonedDateTime dateTime =
+        ZonedDateTime.of(LocalDateTime.now().minusDays(deltaInDays), ZoneId.of("UTC"));
+    return dateTime.format(isoDateTimeFormatter);
+  }
+
+  class PaginatedClouderaYarnApplicationsLoader {
+    private final String host;
+    private final CloseableHttpClient httpClient;
+    private final int limit;
+    private int offset;
+
+    public PaginatedClouderaYarnApplicationsLoader(ClouderaManagerHandle handle) {
+      this.host = handle.getApiURI().toString();
+      this.httpClient = handle.getHttpClient();
+      this.limit = 500;
+      this.offset = 0;
+    }
+
+    public int load(String clusterName, Consumer<List<ApiYARNApplicationDTO>> callback) {
+      return load(clusterName, null, callback);
+    }
+
+    public int load(
+        String clusterName,
+        @Nullable String appType,
+        Consumer<List<ApiYARNApplicationDTO>> callback) {
+      int amountOfLoadedApps = 0;
+      offset = 0;
+      boolean nextLoad = true;
+
+      while (nextLoad) {
+        URI yarnAppsURI = buildYARNApplicationURI(clusterName, appType);
+        List<ApiYARNApplicationDTO> newLoad = load(yarnAppsURI);
+        if (!newLoad.isEmpty()) {
+          callback.accept(newLoad);
+        } else {
+          nextLoad = false;
+        }
+        amountOfLoadedApps += newLoad.size();
+        offset += limit;
+      }
+      return amountOfLoadedApps;
+    }
+
+    private List<ApiYARNApplicationDTO> load(URI yarnAppURI) {
+      try (CloseableHttpResponse resp = httpClient.execute(new HttpGet(yarnAppURI))) {
+        final int statusCode = resp.getStatusLine().getStatusCode();
+        if (!isStatusCodeOK(statusCode)) {
+          throw new RuntimeException(
+              String.format(
+                  "Cloudera Error: YARN application API returned HTTP status %d.", statusCode));
+        }
+        ApiYARNApplicationListDTO yarnAppListDto =
+            parseJsonStreamToObject(resp.getEntity().getContent(), ApiYARNApplicationListDTO.class);
+        return yarnAppListDto.getApplications();
+      } catch (IOException ex) {
+        throw new RuntimeException(ex.getMessage(), ex);
+      }
+    }
+
+    private URI buildYARNApplicationURI(String clusterName, @Nullable String appType) {
```

Comment:
`buildNextPageYARNApplicationURI`, with such name it will be more clear why this method called in a cycle many times with the same parameters.

Reply:
Great idea! I missed it when moved limit/offset parameters into class state. Thanks!